In [125]:
import pandas as pd
import numpy as np
df = pd.read_csv("data/data.csv", sep= ";",encoding="cp1254") 
df.head(5)

,Group,Sex,Age,Patients number per hour,Arrival mode,Injury,Chief_complain,Mental,Pain,NRS_pain,...,BT,Saturation,KTAS_RN,Diagnosis in ED,Disposition,KTAS_expert,Error_group,Length of stay_min,KTAS duration_min,mistriage
0,2,2,71,3,3,2,right ocular pain,1,1,2,...,36.6,100,2,Corneal abrasion,1,4,2,86,"5,00",1
1,1,1,56,12,3,2,right forearm burn,1,1,2,...,36.5,NaN,4,"Burn of hand, firts degree dorsum",1,5,4,64,"3,95",1
2,2,1,68,8,2,2,"arm pain, Lt",1,1,2,...,36.6,98,4,"Fracture of surgical neck of humerus, closed",2,5,4,862,"1,00",1
3,1,2,71,8,1,1,ascites tapping,1,1,3,...,36.5,NaN,4,Alcoholic liver cirrhosis with ascites,1,5,6,108,"9,83",1
4,1,2,58,4,3,1,"distension, abd",1,1,3,...,36.5,NaN,4,Ascites,1,5,8,109,"6,60",1


In [126]:
df.shape

(1267, 24)

In [127]:
df_clean = df.copy() 

    If you use a copy of the original while keeping the original unchanged, then you can refer back to all the untouched data when needed, plus if something goes wrong with copy of the original, another copy can be made from the original which has no changes so its like a fresh start. 

In [128]:
df_clean.info()
df_clean.columns

<class 'pandas.DataFrame'>
RangeIndex: 1267 entries, 0 to 1266
Data columns (total 24 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   Group                     1267 non-null   int64
 1   Sex                       1267 non-null   int64
 2   Age                       1267 non-null   int64
 3   Patients number per hour  1267 non-null   int64
 4   Arrival mode              1267 non-null   int64
 5   Injury                    1267 non-null   int64
 6   Chief_complain            1267 non-null   str  
 7   Mental                    1267 non-null   int64
 8   Pain                      1267 non-null   int64
 9   NRS_pain                  1267 non-null   str  
 10  SBP                       1267 non-null   str  
 11  DBP                       1267 non-null   str  
 12  HR                        1267 non-null   str  
 13  RR                        1267 non-null   str  
 14  BT                        1267 non-null   str  
 15

Index(['Group', 'Sex', 'Age', 'Patients number per hour', 'Arrival mode',
       'Injury', 'Chief_complain', 'Mental', 'Pain', 'NRS_pain', 'SBP', 'DBP',
       'HR', 'RR', 'BT', 'Saturation', 'KTAS_RN', 'Diagnosis in ED',
       'Disposition', 'KTAS_expert', 'Error_group', 'Length of stay_min',
       'KTAS duration_min', 'mistriage'],
      dtype='str')

In [129]:
df_clean.dtypes

Group                       int64
Sex                         int64
Age                         int64
Patients number per hour    int64
Arrival mode                int64
Injury                      int64
Chief_complain                str
Mental                      int64
Pain                        int64
NRS_pain                      str
SBP                           str
DBP                           str
HR                            str
RR                            str
BT                            str
Saturation                    str
KTAS_RN                     int64
Diagnosis in ED               str
Disposition                 int64
KTAS_expert                 int64
Error_group                 int64
Length of stay_min          int64
KTAS duration_min             str
mistriage                   int64
dtype: object

    This data set includes 24 columns for 24 differnt variables taken from a patients record. There are also 1267 rows for 1267 patients who's data has been collected. Vitals signs are using str, probally becuase of missing values and column "titles" are also strings

In [130]:
df_clean.loc[[0,1,2,3,4],("Age","Sex","SBP","DBP","HR","Saturation","KTAS_RN","KTAS_expert")]

,Age,Sex,SBP,DBP,HR,Saturation,KTAS_RN,KTAS_expert
0,71,2,160,100,84,100,2,4
1,56,1,137,75,60,NaN,4,5
2,68,1,130,80,102,98,4,5
3,71,2,139,94,88,NaN,4,5
4,58,2,91,67,93,NaN,4,5


In [131]:
df_clean.iloc[[0,1,2,3,4],[1,2,10,11,12,15,16,19]]

,Sex,Age,SBP,DBP,HR,Saturation,KTAS_RN,KTAS_expert
0,2,71,160,100,84,100,2,4
1,1,56,137,75,60,NaN,4,5
2,1,68,130,80,102,98,4,5
3,2,71,139,94,88,NaN,4,5
4,2,58,91,67,93,NaN,4,5


In [132]:
df_clean = df_clean.rename(columns = {"Saturation":"SpO2","KTAS_RN":"Nurse_KTAS","KTAS_expert":"Expert_KTAS","Chief_complain":"Chief_complaint"})



In [133]:
df_clean.isnull().sum().sort_values(ascending=False)

SpO2                        688
Diagnosis in ED               2
Group                         0
Sex                           0
Arrival mode                  0
Injury                        0
Age                           0
Patients number per hour      0
Mental                        0
Chief_complaint               0
SBP                           0
Pain                          0
DBP                           0
HR                            0
RR                            0
NRS_pain                      0
BT                            0
Nurse_KTAS                    0
Disposition                   0
Expert_KTAS                   0
Error_group                   0
Length of stay_min            0
KTAS duration_min             0
mistriage                     0
dtype: int64

Saturation has the most missing values, with 688 missing values and Diagnosis in ED also has 2 missing values.

In [134]:
pd.set_option('display.max_rows',None)
cols = ["Chief_complaint", "NRS_pain", "SBP","DBP","HR","RR","BT","SpO2"]

for col in cols:
    print(f"\n{col}")
    print(df_clean[col].value_counts())


Chief_complaint
Chief_complaint
abd pain                                 72
dyspnea                                  60
dizziness                                59
fever                                    45
ant. chest pain                          44
Open Wound                               31
headache                                 30
epigastric pain                          22
mental change                            22
general weakness                         21
pain, chest                              16
vomiting                                 16
hematemesis                              14
Left chest pain                          12
Abd. pain                                12
Lt flank pain                            11
syncope                                  10
dysarthria                               10
abdomen pain                             10
palpitation                               9
Vaginal Bleeding                          9
facial injury                             9

??. Chief complaint, SBP, DBP , HR, RR, BT, SPO2
????. Cheif complaint
??? ?? Cheif complaint
?? ??. Cheif complaint
?? ??? Cheif complaint
??? Chief complaint
?? ??? ?? Chief complaint
??? ? ???  Chief complaint 
??? ??? ?? Chief complaint 
#BOŞ!  NRS_pain 

isnull() detects empty values, not non-numeric values so it would not detect things like ?? and #BO$!

In [135]:
df_clean = df_clean.replace(["??","#BOŞ!","????","??? ??","?? ??","?? ???","??? ? ???","???","?? ??? ??","??? ??? ??"],pd.NA)

In [136]:
df_clean.isnull().sum().sort_values(ascending=False)

SpO2                        697
NRS_pain                    556
DBP                          29
SBP                          25
RR                           22
Chief_complaint              21
HR                           20
BT                           18
Diagnosis in ED               2
Mental                        0
Arrival mode                  0
Injury                        0
Sex                           0
Age                           0
Group                         0
Patients number per hour      0
Pain                          0
Nurse_KTAS                    0
Disposition                   0
Expert_KTAS                   0
Error_group                   0
Length of stay_min            0
KTAS duration_min             0
mistriage                     0
dtype: int64

There has been an increase even in catergories that showed no null early becuase we turned the invaild strings into an object that counts as null. 

In [137]:
null=(df_clean.isnull().mean()*100)
print(null[null>0].round(2).sort_values(ascending=False))


SpO2               55.01
NRS_pain           43.88
DBP                 2.29
SBP                 1.97
RR                  1.74
Chief_complaint     1.66
HR                  1.58
BT                  1.42
Diagnosis in ED     0.16
dtype: float64


In [138]:
df_clean[['NRS_pain','SBP','DBP','HR','RR','BT','SpO2','KTAS duration_min']].info(),

<class 'pandas.DataFrame'>
RangeIndex: 1267 entries, 0 to 1266
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   NRS_pain           711 non-null    str  
 1   SBP                1242 non-null   str  
 2   DBP                1238 non-null   str  
 3   HR                 1247 non-null   str  
 4   RR                 1245 non-null   str  
 5   BT                 1249 non-null   str  
 6   SpO2               570 non-null    str  
 7   KTAS duration_min  1267 non-null   str  
dtypes: str(8)
memory usage: 79.3 KB


(None,)

    Because "??", "#BOŞ!" and decimals with commas are strings, the rest of the numbers in that column will also be assigned as string to match those values. 

In [139]:
cols = ["NRS_pain","SBP","DBP","HR","RR","BT","SpO2"] 

for col in cols:
    df_clean[col] = pd.to_numeric(df_clean[col],errors="coerce")

df_clean.describe()

,Group,Sex,Age,Patients number per hour,Arrival mode,Injury,Mental,Pain,NRS_pain,SBP,...,HR,RR,BT,SpO2,Nurse_KTAS,Disposition,Expert_KTAS,Error_group,Length of stay_min,mistriage
count,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000,711.000000,1242.000000,...,1247.000000,1245.000000,1249.000000,570.000000,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000
mean,1.456985,1.521705,54.423836,7.519337,2.820837,1.192581,1.105762,0.563536,4.104079,133.648953,...,83.963111,19.506827,36.580624,97.024561,3.335438,1.609313,3.265983,0.558011,11016.102605,0.250197
std,0.498343,0.499726,19.725033,3.160563,0.807904,0.394482,0.447768,0.496143,1.419332,27.275639,...,16.644096,2.016649,0.545708,4.350556,0.885391,1.157983,0.885803,1.571489,80446.092065,0.628250
min,1.000000,1.000000,16.000000,1.000000,1.000000,1.000000,1.000000,0.000000,1.000000,50.000000,...,32.000000,14.000000,35.000000,20.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000
25%,1.000000,1.000000,37.000000,5.000000,2.000000,1.000000,1.000000,0.000000,3.000000,114.000000,...,72.000000,18.000000,36.200000,97.000000,3.000000,1.000000,3.000000,0.000000,133.000000,0.000000
50%,1.000000,2.000000,57.000000,7.000000,3.000000,1.000000,1.000000,1.000000,4.000000,130.000000,...,82.000000,20.000000,36.500000,98.000000,3.000000,1.000000,3.000000,0.000000,274.000000,0.000000
75%,2.000000,2.000000,71.000000,10.000000,3.000000,1.000000,1.000000,1.000000,5.000000,150.000000,...,96.000000,20.000000,36.800000,98.000000,4.000000,2.000000,4.000000,0.000000,606.500000,0.000000
max,2.000000,2.000000,96.000000,17.000000,7.000000,2.000000,4.000000,1.000000,10.000000,275.000000,...,148.000000,30.000000,41.000000,100.000000,5.000000,7.000000,5.000000,9.000000,709510.000000,2.000000


In [140]:
df_clean.isnull().sum().sort_values(ascending=False)

SpO2                        697
NRS_pain                    556
DBP                          29
SBP                          25
RR                           22
Chief_complaint              21
HR                           20
BT                           18
Diagnosis in ED               2
Mental                        0
Arrival mode                  0
Injury                        0
Sex                           0
Age                           0
Group                         0
Patients number per hour      0
Pain                          0
Nurse_KTAS                    0
Disposition                   0
Expert_KTAS                   0
Error_group                   0
Length of stay_min            0
KTAS duration_min             0
mistriage                     0
dtype: int64

In [141]:
df_clean.dtypes

Group                         int64
Sex                           int64
Age                           int64
Patients number per hour      int64
Arrival mode                  int64
Injury                        int64
Chief_complaint                 str
Mental                        int64
Pain                          int64
NRS_pain                    float64
SBP                         float64
DBP                         float64
HR                          float64
RR                          float64
BT                          float64
SpO2                        float64
Nurse_KTAS                    int64
Diagnosis in ED                 str
Disposition                   int64
Expert_KTAS                   int64
Error_group                   int64
Length of stay_min            int64
KTAS duration_min               str
mistriage                     int64
dtype: object

In [142]:
df_clean["KTAS duration_min"] = df_clean["KTAS duration_min"].astype(str).str.replace(",", ".", regex=False)
df_clean["KTAS duration_min"] = pd.to_numeric(df_clean["KTAS duration_min"],errors="coerce")
float(df_clean["KTAS duration_min"].max())

17.37

In [143]:
float(df_clean["KTAS duration_min"].min())

1.0

In [144]:
float((df_clean["KTAS duration_min"].mean()).round(3))

5.497

In [145]:
float(df_clean["KTAS duration_min"].median())

4.28

In [146]:
df_clean.shape

(1267, 24)

In [147]:
df_clean.dtypes


Group                         int64
Sex                           int64
Age                           int64
Patients number per hour      int64
Arrival mode                  int64
Injury                        int64
Chief_complaint                 str
Mental                        int64
Pain                          int64
NRS_pain                    float64
SBP                         float64
DBP                         float64
HR                          float64
RR                          float64
BT                          float64
SpO2                        float64
Nurse_KTAS                    int64
Diagnosis in ED                 str
Disposition                   int64
Expert_KTAS                   int64
Error_group                   int64
Length of stay_min            int64
KTAS duration_min           float64
mistriage                     int64
dtype: object

In [148]:
df_clean.isnull().sum()


Group                         0
Sex                           0
Age                           0
Patients number per hour      0
Arrival mode                  0
Injury                        0
Chief_complaint              21
Mental                        0
Pain                          0
NRS_pain                    556
SBP                          25
DBP                          29
HR                           20
RR                           22
BT                           18
SpO2                        697
Nurse_KTAS                    0
Diagnosis in ED               2
Disposition                   0
Expert_KTAS                   0
Error_group                   0
Length of stay_min            0
KTAS duration_min             0
mistriage                     0
dtype: int64

In [149]:
df_clean.columns

Index(['Group', 'Sex', 'Age', 'Patients number per hour', 'Arrival mode',
       'Injury', 'Chief_complaint', 'Mental', 'Pain', 'NRS_pain', 'SBP', 'DBP',
       'HR', 'RR', 'BT', 'SpO2', 'Nurse_KTAS', 'Diagnosis in ED',
       'Disposition', 'Expert_KTAS', 'Error_group', 'Length of stay_min',
       'KTAS duration_min', 'mistriage'],
      dtype='str')

In [150]:
df.columns

Index(['Group', 'Sex', 'Age', 'Patients number per hour', 'Arrival mode',
       'Injury', 'Chief_complain', 'Mental', 'Pain', 'NRS_pain', 'SBP', 'DBP',
       'HR', 'RR', 'BT', 'Saturation', 'KTAS_RN', 'Diagnosis in ED',
       'Disposition', 'KTAS_expert', 'Error_group', 'Length of stay_min',
       'KTAS duration_min', 'mistriage'],
      dtype='str')

In [151]:
df.shape

(1267, 24)

In [152]:
df.dtypes

Group                       int64
Sex                         int64
Age                         int64
Patients number per hour    int64
Arrival mode                int64
Injury                      int64
Chief_complain                str
Mental                      int64
Pain                        int64
NRS_pain                      str
SBP                           str
DBP                           str
HR                            str
RR                            str
BT                            str
Saturation                    str
KTAS_RN                     int64
Diagnosis in ED               str
Disposition                 int64
KTAS_expert                 int64
Error_group                 int64
Length of stay_min          int64
KTAS duration_min             str
mistriage                   int64
dtype: object

In [153]:
df.isnull().sum()

Group                         0
Sex                           0
Age                           0
Patients number per hour      0
Arrival mode                  0
Injury                        0
Chief_complain                0
Mental                        0
Pain                          0
NRS_pain                      0
SBP                           0
DBP                           0
HR                            0
RR                            0
BT                            0
Saturation                  688
KTAS_RN                       0
Diagnosis in ED               2
Disposition                   0
KTAS_expert                   0
Error_group                   0
Length of stay_min            0
KTAS duration_min             0
mistriage                     0
dtype: int64

1. How many rows and columns are in the dataset?
    1.There are 1267 rows and 24 columns, 1267 for 1267 patient records and 24 variables
2. Which column has the most standard missing values?
    2.Saturation, also known as SpO2 in df_clean with 688 missing values, and 697 after turning invalid input into null.
3. Which special values represent missing data?
    3.NaN.
4. Why did some numeric columns initially have the `object` data type?
    4.Some of those values were actually strings making the rest of the column of string type
5. How were the vital-sign columns converted into numeric columns?
    5.Using  pd.to_numeric(df_clean[col],errors="coerce") which would turn the values from string type to float type.
6. How was `KTAS duration_min` converted into numeric data?
    6.Since some of the numbers used a comma instead of a period for decimals, I first replaced the commas with a period to ensure that they would register as a number when converted and not a NaN. Then use pd.to_numeric(df_clean["KTAS duration min"]).
7. Which column was the most difficult to clean?
    7.The Chief complaint column becuase it contained differnt combiantion of ??
8. Why is it useful to preserve the original DataFrame?
    8.If you use a copy of the original while keeping the original unchanged, then you can refer back to all the untouched data when needed, plus if something goes wrong with copy of the original, another copy can be made from the original which has no changes so its like a fresh start.  
    

In [154]:
bins = [0,19,39,59,79,np.inf]
ages = ['AGE_19','AGE_39','AGE_59','AGE_79','AGE_80+']
Ages = pd.cut(df_clean['Age'],bins=bins, labels= ages, include_lowest=True)

df_clean = df_clean.drop(columns=ages, errors='ignore')
age_dummies = pd.get_dummies(Ages, dtype=int)
loc = df_clean.columns.get_loc('Age') + 1

for i, col in enumerate(age_dummies.columns):
    df_clean.insert(loc + i, col, age_dummies[col])
df_clean.head(4)

,Group,Sex,Age,AGE_19,AGE_39,AGE_59,AGE_79,AGE_80+,Patients number per hour,Arrival mode,...,BT,SpO2,Nurse_KTAS,Diagnosis in ED,Disposition,Expert_KTAS,Error_group,Length of stay_min,KTAS duration_min,mistriage
0,2,2,71,0,0,0,1,0,3,3,...,36.6,100.0,2,Corneal abrasion,1,4,2,86,5.00,1
1,1,1,56,0,0,1,0,0,12,3,...,36.5,NaN,4,"Burn of hand, firts degree dorsum",1,5,4,64,3.95,1
2,2,1,68,0,0,0,1,0,8,2,...,36.6,98.0,4,"Fracture of surgical neck of humerus, closed",2,5,4,862,1.00,1
3,1,2,71,0,0,0,1,0,8,1,...,36.5,NaN,4,Alcoholic liver cirrhosis with ascites,1,5,6,108,9.83,1


In [155]:
age_group = pd.cut(df_clean['Age'], bins= bins, include_lowest=True).value_counts().sort_index()
print(age_group)

Age
(-0.001, 19.0]     34
(19.0, 39.0]      309
(39.0, 59.0]      371
(59.0, 79.0]      437
(79.0, inf]       116
Name: count, dtype: int64


In [156]:
print(df_clean['Sex'].unique())
sex_map = {
    1: 'SEX_Female',
    2: 'SEX_Male'
    
}
sex = df_clean['Sex'].map(sex_map)
sex_dummies = pd.get_dummies(sex, dtype=int)
loc = df_clean.columns.get_loc('Sex') + 1

for i, col in enumerate(sex_dummies.columns):
    df_clean.insert(loc + i, col, sex_dummies[col])

df_clean.head(4)


[2 1]


,Group,Sex,SEX_Female,SEX_Male,Age,AGE_19,AGE_39,AGE_59,AGE_79,AGE_80+,...,BT,SpO2,Nurse_KTAS,Diagnosis in ED,Disposition,Expert_KTAS,Error_group,Length of stay_min,KTAS duration_min,mistriage
0,2,2,0,1,71,0,0,0,1,0,...,36.6,100.0,2,Corneal abrasion,1,4,2,86,5.00,1
1,1,1,1,0,56,0,0,1,0,0,...,36.5,NaN,4,"Burn of hand, firts degree dorsum",1,5,4,64,3.95,1
2,2,1,1,0,68,0,0,0,1,0,...,36.6,98.0,4,"Fracture of surgical neck of humerus, closed",2,5,4,862,1.00,1
3,1,2,0,1,71,0,0,0,1,0,...,36.5,NaN,4,Alcoholic liver cirrhosis with ascites,1,5,6,108,9.83,1


In [157]:
df_clean['SEX_Female'].isnull().value_counts(dropna=False)

SEX_Female
False    1267
Name: count, dtype: int64

In [158]:
df_clean['SEX_Male'].isnull().value_counts(dropna=False)

SEX_Male
False    1267
Name: count, dtype: int64

In [159]:
df_clean.tail(4)

,Group,Sex,SEX_Female,SEX_Male,Age,AGE_19,AGE_39,AGE_59,AGE_79,AGE_80+,...,BT,SpO2,Nurse_KTAS,Diagnosis in ED,Disposition,Expert_KTAS,Error_group,Length of stay_min,KTAS duration_min,mistriage
1263,2,2,0,1,81,0,0,0,0,1,...,36.4,97.0,4,Dysuria,1,4,0,1000,2.0,0
1264,2,2,0,1,81,0,0,0,0,1,...,36.2,99.0,3,Dizziness and giddiness,1,3,0,310,4.0,0
1265,2,1,1,0,81,0,0,0,0,1,...,36.6,98.0,3,"Cord compression, unspecified",7,3,0,475,5.0,0
1266,2,2,0,1,87,0,0,0,0,1,...,36.2,98.0,3,Dyspnoea,2,3,0,887,2.0,0


In [160]:
senior = np.where(df_clean['Age'] >= 65,'Yes','No')
loc = df_clean.columns.get_loc('Age') + 1
df_clean.insert(loc, 'Senior',senior)
df_clean.head(2)

,Group,Sex,SEX_Female,SEX_Male,Age,Senior,AGE_19,AGE_39,AGE_59,AGE_79,...,BT,SpO2,Nurse_KTAS,Diagnosis in ED,Disposition,Expert_KTAS,Error_group,Length of stay_min,KTAS duration_min,mistriage
0,2,2,0,1,71,Yes,0,0,0,1,...,36.6,100.0,2,Corneal abrasion,1,4,2,86,5.00,1
1,1,1,1,0,56,No,0,0,1,0,...,36.5,NaN,4,"Burn of hand, firts degree dorsum",1,5,4,64,3.95,1


In [161]:
(df_clean['Age'] >= 65).map({True: 'Senior', False: 'Non-Senior'}).value_counts().sort_index()

Age
Non-Senior    800
Senior        467
Name: count, dtype: int64

In [162]:
ture = df_clean['SpO2'].notna()
low_Spo2 = df_clean['SpO2'].lt(95).where(ture).astype("boolean")
loc = df_clean.columns.get_loc('SpO2') + 1
df_clean.insert(loc, 'Low_SpO2',low_Spo2)



In [163]:
low_Spo2 = df_clean['SpO2'].lt(95).where(ture,'Missing').value_counts()
print(low_Spo2)

SpO2
Missing    697
False      523
True        47
Name: count, dtype: int64


Explain why missing SpO₂ values must be kept separate from False.

They are missing, so we dont know what the true value is. Since they could be true or flase, making them false instead of excluding them could make the false to true rate significantly higher than it actually is. 

In [184]:
#df_clean[df_clean['mistriage'] == 0] #correct
#df_clean[df_clean['mistriage'] == 1] #over-triage
#df_clean[df_clean['mistriage'] == 2] #under-triage

In [165]:
KTAS_bool = df_clean['KTAS_match'] = df_clean['Expert_KTAS'] == df_clean['Nurse_KTAS']
ktas_match = df_clean.pop("KTAS_match")
loc = df_clean.columns.get_loc('Expert_KTAS') + 1
df_clean.insert(loc,"KTAS_match",ktas_match)
df_clean.head(5)


,Group,Sex,SEX_Female,SEX_Male,Age,Senior,AGE_19,AGE_39,AGE_59,AGE_79,...,Low_SpO2,Nurse_KTAS,Diagnosis in ED,Disposition,Expert_KTAS,KTAS_match,Error_group,Length of stay_min,KTAS duration_min,mistriage
0,2,2,0,1,71,Yes,0,0,0,1,...,False,2,Corneal abrasion,1,4,False,2,86,5.00,1
1,1,1,1,0,56,No,0,0,1,0,...,<NA>,4,"Burn of hand, firts degree dorsum",1,5,False,4,64,3.95,1
2,2,1,1,0,68,Yes,0,0,0,1,...,False,4,"Fracture of surgical neck of humerus, closed",2,5,False,4,862,1.00,1
3,1,2,0,1,71,Yes,0,0,0,1,...,<NA>,4,Alcoholic liver cirrhosis with ascites,1,5,False,6,108,9.83,1
4,1,2,0,1,58,No,0,0,1,0,...,<NA>,4,Ascites,1,5,False,8,109,6.60,1


In [166]:
print("Matching:", KTAS_bool.sum())
print("Non-matching:", 1267-KTAS_bool.sum())

Matching: 1081
Non-matching: 186


In [167]:
pd.crosstab(df_clean['mistriage'],df_clean['KTAS_match'])

KTAS_match,False,True
mistriage,,
0,0,1081
1,55,0
2,131,0


1.In mistriage there is 1 for overtriage, 2 for under and 0 for a correct score.
2.It contains both under, over, and correct match info
3.It does so yea
4.If someone wants to make new catergories, they can be sure to include, under and over, triage, not just mistriage

In [168]:
df_clean2 = df_clean.copy()

In [169]:
df_clean2.duplicated()

0       False
1       False
2       False
3       False
4       False
5       False
6       False
7       False
8       False
9       False
10      False
11      False
12      False
13      False
14      False
15      False
16      False
17      False
18      False
19      False
20      False
21      False
22      False
23      False
24      False
25      False
26      False
27      False
28      False
29      False
30      False
31      False
32      False
33      False
34      False
35      False
36      False
37      False
38      False
39      False
40      False
41      False
42      False
43      False
44      False
45      False
46      False
47      False
48      False
49      False
50      False
51      False
52      False
53      False
54      False
55      False
56      False
57      False
58      False
59      False
60      False
61      False
62      False
63      False
64      False
65      False
66      False
67      False
68      False
69      False
70      False
71    

In [170]:
df_clean2.sum()

Group                                                                    1846
Sex                                                                      1928
SEX_Female                                                                606
SEX_Male                                                                  661
Age                                                                     68955
Senior                      YesNoYesYesNoNoNoYesNoNoNoNoNoNoNoYesYesNoNoNo...
AGE_19                                                                     34
AGE_39                                                                    309
AGE_59                                                                    371
AGE_79                                                                    437
AGE_80+                                                                   116
Patients number per hour                                                 9527
Arrival mode                                                    

In [171]:
df_clean2.drop_duplicates()
df_clean2.head()

,Group,Sex,SEX_Female,SEX_Male,Age,Senior,AGE_19,AGE_39,AGE_59,AGE_79,...,Low_SpO2,Nurse_KTAS,Diagnosis in ED,Disposition,Expert_KTAS,KTAS_match,Error_group,Length of stay_min,KTAS duration_min,mistriage
0,2,2,0,1,71,Yes,0,0,0,1,...,False,2,Corneal abrasion,1,4,False,2,86,5.00,1
1,1,1,1,0,56,No,0,0,1,0,...,<NA>,4,"Burn of hand, firts degree dorsum",1,5,False,4,64,3.95,1
2,2,1,1,0,68,Yes,0,0,0,1,...,False,4,"Fracture of surgical neck of humerus, closed",2,5,False,4,862,1.00,1
3,1,2,0,1,71,Yes,0,0,0,1,...,<NA>,4,Alcoholic liver cirrhosis with ascites,1,5,False,6,108,9.83,1
4,1,2,0,1,58,No,0,0,1,0,...,<NA>,4,Ascites,1,5,False,8,109,6.60,1


In [172]:
df_clean2.shape

(1267, 34)

Just in case it is another patient that was very similar or it was there for a reason, i.e. an example case
There were no duplicates, but its importnat to check anyway just in case before using the data set and giving it to other scientists or like ai or smth. 

In [173]:
df_clean.loc[(df_clean["Senior"] == "Yes") & (df_clean["Low_SpO2"]) & (df_clean["HR"] >= 100),["Age","SEX_Female","SEX_Male","DBP","HR","SpO2","Nurse_KTAS","Expert_KTAS"]] 

,Age,SEX_Female,SEX_Male,DBP,HR,SpO2,Nurse_KTAS,Expert_KTAS
132,82,0,1,100.0,116.0,68.0,2,1
134,92,1,0,90.0,110.0,88.0,3,2
153,65,0,1,NaN,100.0,94.0,2,1
405,72,1,0,60.0,106.0,93.0,2,2
412,74,1,0,60.0,110.0,91.0,3,3
434,80,0,1,80.0,118.0,76.0,2,2
550,81,1,0,60.0,110.0,93.0,3,3
1093,70,0,1,80.0,100.0,90.0,3,3
1147,77,0,1,80.0,106.0,93.0,3,3
1169,80,1,0,70.0,110.0,93.0,3,3


In [174]:
mask = (df_clean["Senior"] == "Yes") & (df_clean["Low_SpO2"]) & (df_clean["HR"] >= 100)
mask.sum()

np.int64(11)

In [175]:
df_clean.query('Senior == "Yes" and Low_SpO2 and HR >= 100')[["Age", "SEX_Female","SEX_Male","DBP","HR","SpO2","Nurse_KTAS","Expert_KTAS"]]

,Age,SEX_Female,SEX_Male,DBP,HR,SpO2,Nurse_KTAS,Expert_KTAS
132,82,0,1,100.0,116.0,68.0,2,1
134,92,1,0,90.0,110.0,88.0,3,2
153,65,0,1,NaN,100.0,94.0,2,1
405,72,1,0,60.0,106.0,93.0,2,2
412,74,1,0,60.0,110.0,91.0,3,3
434,80,0,1,80.0,118.0,76.0,2,2
550,81,1,0,60.0,110.0,93.0,3,3
1093,70,0,1,80.0,100.0,90.0,3,3
1147,77,0,1,80.0,106.0,93.0,3,3
1169,80,1,0,70.0,110.0,93.0,3,3


In [176]:
print(df_clean.query('Senior == "Yes" and Low_SpO2 and HR >= 100')[["Age", "SEX_Female","SEX_Male","DBP","HR","SpO2","Nurse_KTAS","Expert_KTAS"]].reset_index().equals(df_clean.loc[(df_clean["Senior"] == "Yes") & (df_clean["Low_SpO2"]) & (df_clean["HR"] >= 100),["Age","SEX_Female","SEX_Male","DBP","HR","SpO2","Nurse_KTAS","Expert_KTAS"]].reset_index()))

True


Both methods returned the same records and we know that becuase without reset_index they are equal and when the both use reset_index they are equal

In [177]:
df_clean.groupby('Group')['SpO2'].apply(lambda x: x.isnull().sum())

Group
1    688
2      9
Name: SpO2, dtype: int64

In [178]:
df_clean.groupby('Group')['SpO2'].apply(lambda x: x.isnull().mean()*100)

Group
1    100.000000
2      1.554404
Name: SpO2, dtype: float64

Group 1.
Very skewed towards group 1.
Maybe, since all local didnt have SpO2


In [179]:
analysis_df = df_clean.copy() 
analysis_df.drop(columns=['Sex','Patients number per hour','Diagnosis in ED','mistriage','Error_group','Low_SpO2','Senior'], inplace=True)
analysis_df.head(5)


,Group,SEX_Female,SEX_Male,Age,AGE_19,AGE_39,AGE_59,AGE_79,AGE_80+,Arrival mode,...,HR,RR,BT,SpO2,Nurse_KTAS,Disposition,Expert_KTAS,KTAS_match,Length of stay_min,KTAS duration_min
0,2,0,1,71,0,0,0,1,0,3,...,84.0,18.0,36.6,100.0,2,1,4,False,86,5.00
1,1,1,0,56,0,0,1,0,0,3,...,60.0,20.0,36.5,NaN,4,1,5,False,64,3.95
2,2,1,0,68,0,0,0,1,0,2,...,102.0,20.0,36.6,98.0,4,2,5,False,862,1.00
3,1,0,1,71,0,0,0,1,0,1,...,88.0,20.0,36.5,NaN,4,1,5,False,108,9.83
4,1,0,1,58,0,0,1,0,0,3,...,93.0,18.0,36.5,NaN,4,1,5,False,109,6.60


In [186]:
analysis_df.shape

(1267, 27)

In [187]:
analysis_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1267 entries, 0 to 1266
Data columns (total 27 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Group               1267 non-null   int64  
 1   SEX_Female          1267 non-null   int64  
 2   SEX_Male            1267 non-null   int64  
 3   Age                 1267 non-null   int64  
 4   AGE_19              1267 non-null   int64  
 5   AGE_39              1267 non-null   int64  
 6   AGE_59              1267 non-null   int64  
 7   AGE_79              1267 non-null   int64  
 8   AGE_80+             1267 non-null   int64  
 9   Arrival mode        1267 non-null   int64  
 10  Injury              1267 non-null   int64  
 11  Chief_complaint     1246 non-null   str    
 12  Mental              1267 non-null   int64  
 13  Pain                1267 non-null   int64  
 14  NRS_pain            711 non-null    float64
 15  SBP                 1242 non-null   float64
 16  DBP              

In [181]:
analysis_df.isnull().sum()

Group                   0
SEX_Female              0
SEX_Male                0
Age                     0
AGE_19                  0
AGE_39                  0
AGE_59                  0
AGE_79                  0
AGE_80+                 0
Arrival mode            0
Injury                  0
Chief_complaint        21
Mental                  0
Pain                    0
NRS_pain              556
SBP                    25
DBP                    29
HR                     20
RR                     22
BT                     18
SpO2                  697
Nurse_KTAS              0
Disposition             0
Expert_KTAS             0
KTAS_match              0
Length of stay_min      0
KTAS duration_min       0
dtype: int64

In [182]:
analysis_df.describe()

,Group,SEX_Female,SEX_Male,Age,AGE_19,AGE_39,AGE_59,AGE_79,AGE_80+,Arrival mode,...,DBP,HR,RR,BT,SpO2,Nurse_KTAS,Disposition,Expert_KTAS,Length of stay_min,KTAS duration_min
count,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000,...,1238.000000,1247.000000,1245.000000,1249.000000,570.000000,1267.000000,1267.000000,1267.000000,1267.000000,1267.000000
mean,1.456985,0.478295,0.521705,54.423836,0.026835,0.243883,0.292818,0.344909,0.091555,2.820837,...,79.780291,83.963111,19.506827,36.580624,97.024561,3.335438,1.609313,3.265983,11016.102605,5.497466
std,0.498343,0.499726,0.499726,19.725033,0.161665,0.429593,0.455235,0.475526,0.288510,0.807904,...,15.154292,16.644096,2.016649,0.545708,4.350556,0.885391,1.157983,0.885803,80446.092065,3.459470
min,1.000000,0.000000,0.000000,16.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,...,31.000000,32.000000,14.000000,35.000000,20.000000,1.000000,1.000000,1.000000,0.000000,1.000000
25%,1.000000,0.000000,0.000000,37.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.000000,...,70.000000,72.000000,18.000000,36.200000,97.000000,3.000000,1.000000,3.000000,133.000000,3.000000
50%,1.000000,0.000000,1.000000,57.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.000000,...,80.000000,82.000000,20.000000,36.500000,98.000000,3.000000,1.000000,3.000000,274.000000,4.280000
75%,2.000000,1.000000,1.000000,71.000000,0.000000,0.000000,1.000000,1.000000,0.000000,3.000000,...,90.000000,96.000000,20.000000,36.800000,98.000000,4.000000,2.000000,4.000000,606.500000,7.000000
max,2.000000,1.000000,1.000000,96.000000,1.000000,1.000000,1.000000,1.000000,1.000000,7.000000,...,160.000000,148.000000,30.000000,41.000000,100.000000,5.000000,7.000000,5.000000,709510.000000,17.370000


In [183]:
analysis_df.to_csv('emergency_triage_clean.csv',index=False)

1. Which age group contains the most patients?
    The 60-79 age group had the most patients with 437 patients.
2. How many patients are aged 65 or older?
    There were 467 senior patients.
3. How many patients have SpO₂ below 95?
    There were 47 patients with a SpO2 below 95.
4. How many SpO₂ values are missing?
    There were 697 missing SpO2 values.
5. Why were missing SpO₂ values not classified as normal?
    They could be normal, but they could also be abnormal, so to make them all normal might skew the percentage of normal and abnormal SpO2 levels.
6. What percentage of nurse and expert KTAS values match?
    About 85%.
7. Were any duplicate rows found?
    There were no duplicate rows found.
8. Did Boolean indexing and `query()` produce the same result?
    They produced the same  result. 
9. Were missing SpO₂ values evenly distributed across `Group`?
    No, 100% of the SpO2 values in group 1 was missing while 1.6% of group 2 had missing SpO2 values.
10. What variables were included in `analysis_df`?
    Group,Age,Age_group,Sex_name,Arrival mode,Injury,Chief_complaint,Mental,Pain,NRS_pain,SBP,DBP,HR,RR,BT,SpO2,Nurse_KTAS,Expert_KTAS,KTAS_match,KTAS duration_min,Length of stay_min,Disposition
11. Which Pandas function was the most difficult to use?
    groupby, cause you also need to use apply and lambda and I didn't know that so that was a bit tricky.